udtf.go

```diff
@@ -0,0 +1,493 @@
+
+	tfunc := h.Value().(pinnedValue[tableFunction[T]]).value
+	config := tfunc.Config
+
+	argCount := len(config.Arguments)
+	args := make([]any, argCount)
+	namedArgs := make(map[string]any)
+	for i, t := range config.Arguments {
+		value := C.duckdb_bind_get_parameter(info, C.idx_t(i))
+		var err error
+		args[i], err = getValue(t, value)
+		if err != nil {
+			errstr := C.CString(err.Error())
+			defer C.free(unsafe.Pointer(errstr))
+			C.duckdb_bind_set_error(info, errstr)
+			return
+		}
+	}
+
+	for name, t := range config.NamedArguments {
+		argName := C.CString(name)
+		defer C.free(unsafe.Pointer(argName))
+		value := C.duckdb_bind_get_named_parameter(info, argName)
+
+		var err error
+		namedArgs[name], err = getValue(t, value)
+		if err != nil {
+			errstr := C.CString(err.Error())
+			defer C.free(unsafe.Pointer(errstr))
+			C.duckdb_bind_set_error(info, errstr)
+			return
+		}
+	}
+
+	instance, err := tfunc.BindArguments(namedArgs, args...)
+	if err != nil {
+		errstr := C.CString(err.Error())
+		defer C.free(unsafe.Pointer(errstr))
+		C.duckdb_bind_set_error(info, errstr)
+		return
+	}
+
+	columns := instance.Columns()
+
+	instanceData := tableFunctionData{
+		fun:        instance,
+		projection: make([]int, len(columns)),
+	}
+
+	for i, v := range columns {
+		dt := v.T.logicalType()
+		defer C.duckdb_destroy_logical_type(&dt)
+		colName := C.CString(v.Name)
+		defer C.free(unsafe.Pointer(colName))
+		C.duckdb_bind_add_result_column(info, colName, dt)
+
+		instanceData.projection[i] = -1
+	}
+
+	cardinality := instance.Cardinality()
+	if cardinality != nil {
+		C.duckdb_bind_set_cardinality(info, C.idx_t(cardinality.Cardinality), C.bool(cardinality.IsExact))
+	}
+
+	pinnedInstanceData := pinnedValue[tableFunctionData]{
+		pinner: &runtime.Pinner{},
+		value:  instanceData,
+	}
+
+	handle := cgo.NewHandle(pinnedInstanceData)
+	pinnedInstanceData.pinner.Pin(&handle)
+	C.duckdb_bind_set_bind_data(info, unsafe.Pointer(&handle), C.duckdb_delete_callback_t(C.udf_destroy_data))
+}
+
+//export udf_init
+func udf_init(info C.duckdb_init_info) {
+	h := *(*cgo.Handle)(C.duckdb_init_get_bind_data(info))
+	instance := h.Value().(pinnedValue[tableFunctionData]).value
+
+	columnCount := C.duckdb_init_get_column_count(info)
+	for i := C.idx_t(0); i < columnCount; i++ {
+		srcPos := int(C.duckdb_init_get_column_index(info, i))
+		instance.projection[srcPos] = int(i)
+	}
+	instance.fun.(tableSource).Init()
+}
+
+//export udf_init_threaded
+func udf_init_threaded(info C.duckdb_init_info) {
+	h := *(*cgo.Handle)(C.duckdb_init_get_bind_data(info))
+	instance := h.Value().(pinnedValue[tableFunctionData]).value
+
+	columnCount := C.duckdb_init_get_column_count(info)
+	for i := C.idx_t(0); i < columnCount; i++ {
+		srcPos := int(C.duckdb_init_get_column_index(info, i))
+		instance.projection[srcPos] = int(i)
+	}
+
+	initData := instance.fun.(threadedTableSource).Init()
+	C.duckdb_init_set_max_threads(info, C.idx_t(initData.MaxThreads))
+}
```

Comment:
Add a function for the following lines to avoid duplicate code.
```Go
	h := *(*cgo.Handle)(C.duckdb_init_get_bind_data(info))
 	instance := h.Value().(pinnedValue[tableFunctionData]).value

 	columnCount := C.duckdb_init_get_column_count(info)
 	for i := C.idx_t(0); i < columnCount; i++ {
 		srcPos := int(C.duckdb_init_get_column_index(info, i))
 		instance.projection[srcPos] = int(i)
 	}
```